experimental: ServiceCache/ServiceManager adapters, LocalRedirectPolicy support

joamaki opened the pull request:
This implements adapters that bridge pkg/loadbalancer/experimental to the `ServiceCache` and `ServiceManager` APIs and brings in LocalRedirectPolicy support for the experimental control-plane. The CiliumEnvoyConfig handling for the new control-plane was rewritten to support headless services (the original approach of watching Table[Frontend] of course didn't work as headless services have no frontends...).

The PR is large as it contains many fixes uncovered when running the full CI with the new control-plane (https://github.com/cilium/cilium/pull/37083) and iterating on fixing issues. Since the changes are within the experimental packages I've decided to just do a large PR to push them in.

brb committed:
ginkgo: Remove preflight check for |LBMAP| == |service list|

The check no longer makes sense, as "cilium service list" no longer
returns all service frontends, while "cilium bpf lb list" does.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>
Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Implement handling of LocalRedirectPolicies

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Headless services rewrite

In order to support referencing headless services the CEC controller
cannot be built around watching changes on the Table[Frontend].
Reimplement the controller loop to not use Changes() but rather do
queries of the relevant tables and use WatchSet to know when to recompute.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
redirectpolicy,skiplbmap: Add OpenOrCreate, AllLB*, DeleteLB*

To allow the SkipLBMap to be separately constructed, opened
and closed add OpenOrCreate() and Close().

For the follow-up new local redirect policy implementation add the
AllLB* iterators and the DeleteLB* methods.

In pkg/redirectpolicy, do nothing if the experimental control-plane
is enabled (as that will replace this implementation).

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Move HaveNetNSCookieSupported to experimental package

We need this also in reflector.go.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental/redirectpolicy: Implement REST API handler

Replace the REST API handler of the old implementation when enabled.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
gh/workflows: Create LB IPAM pool in conformance-k8s tests

Otherwise, the "[sig-network] Services should fail health check node
port if there are only terminating endpoints" fails.

In the exp/lb we create HealthCheckNodePort only after a corresponding
LoadBalancer service got assigned VIP. Previously, Cilium was unable to
allocate LB VIPs.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>
Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb commented on pkg/loadbalancer/experimental/adapters.go:
```diff
@@ -0,0 +1,682 @@
+
+// newAdapters constructs the ServiceCache and ServiceManager adapters. This is separate
+// from [decorateAdapters] in order to have access to the module's job.Group which is not
+// accessibel in the [cell.DecorateAll] scope.
```
nit: accesible

brb commented on pkg/loadbalancer/experimental/adapters.go:
```diff
@@ -0,0 +1,682 @@
+
+		for name := range changed {
+			n, stateFound := state[name]
+			if svc, _, found := s.services.Get(txn, ServiceByName(name)); found {
```
A code comment explaining why we do the get from svc tables and under what circumstances `found` might be `false` would be helpful.

dylandreimerink commented on pkg/loadbalancer/experimental/adapters.go:
```diff
@@ -0,0 +1,682 @@
+	name := loadbalancer.ServiceName{
+		Namespace: service.Namespace,
+		Name:      service.Name,
+		// TODO ClusterAware
+		// Cluster: csvc.Cluster,
```
Did we miss something that should be in the PR? If not, we should turn this into an issue and add the link here so things are properly tracked and context is preserved.

dylandreimerink commented on pkg/loadbalancer/experimental/adapters.go:
```diff
@@ -0,0 +1,682 @@
+	name := loadbalancer.ServiceName{
+		Namespace: service.Namespace,
+		Name:      service.Name,
+		// TODO ClusterAware
+		// Cluster: csvc.Cluster,
```
See other comment on TODOs

dylandreimerink commented on pkg/loadbalancer/experimental/adapters.go:
```diff
@@ -0,0 +1,682 @@
+			Annotations: fe.service.Annotations,
+			Backends:    bes,
+
+			ForwardingMode:            "", // FIXME (not implemented)
```
Same as with the TODOs. I understand not implementing everything at once, but it would be good to track all of that pending work.

joamaki replied:
These are all tracked as tasks in https://github.com/cilium/cilium/issues/34176.

dylandreimerink commented on pkg/loadbalancer/experimental/healthserver.go:
```diff
@@ -71,14 +75,21 @@ func registerHealthServer(params healthServerParams) {
 		portByService: map[lb.ServiceName]uint16{},
 	}
+
+	// TODO(brb) Handle ipv6 + cluster-id
```
Issue please

joamaki replied:
@brb do you have the context for this? What's needed?

joamaki replied:
Removed comment and added to task list to add IPv6 test case for health server.

jrajahalme reviewed (approved):
Could not follow all the detail, but LGTM :-)

dylandreimerink commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -241,82 +241,115 @@ var (
+			continue
+		}
+		// TODO(brb) warn logging?
+		m.Unpin()
```
You can simply to a [`os.Remove`](https://pkg.go.dev/os#Remove) since pins are pseudo files. I would check the error, if the file doesn't exists, ignore it. If its some other error (permission denied for example) then log it.

joamaki replied:
@brb shall we just drop the TODO? or do what @dylandreimerink suggested.

joamaki replied:
`Unpin()` will return nil when file does not exist. I'll just check the error and always warn.

dylandreimerink commented on pkg/loadbalancer/experimental/node_addr_reconciler.go:
```diff
@@ -51,11 +53,15 @@ type nodePortAddrReconciler struct {
 }
 
 func (r *nodePortAddrReconciler) nodePortAddressReconcilerLoop(ctx context.Context, health cell.Health) error {
+	limiter := rate.NewLimiter(time.Second, 1)
```
Nit: can we add a comment, how did we get this number? Even if its out of somebodies ass, always good to know if its important or can be freely changed.

joamaki replied:
Added.

dylandreimerink commented on pkg/loadbalancer/experimental/reflector.go:
```diff
@@ -125,12 +131,13 @@ func runResourceReflector(ctx context.Context, p reflectorParams, initComplete f
 		}
 		if err := p.Writer.UpsertServiceAndFrontends(txn, svc, fes...); err != nil {
-			panic("BUG: UpsertServiceAndFrontends: " + err.Error())
+			p.Log.Error("BUG: Unexpected failure in UpsertServiceAndFrontends",
```
No panic is good. But side effect is that users will not be notified anymore. Can we propagate the job health and degrade it in such instances? Seems like the logical step down from panic'ing

joamaki replied:
Ah good idea! Especially as now there is actual potential failure here (frontend address conflict).

joamaki replied:
Added the handling to update health when any upsert fails. Can't really handle the deletions as those would just stick around forever and currently there is no expected errors that can happen from deletions, so still logging there instead.

dylandreimerink commented on pkg/ciliumenvoyconfig/exp_controller.go:
```diff
@@ -257,87 +113,253 @@ func (c *cecController) nodeLabelController(ctx context.Context, health cell.Hea
+	for be := range backends {
+		inst := be.GetInstance(serviceName)
+		if inst.State != loadbalancer.BackendStateActive {
+			// FIXME: should use the control-plane logic for picking the backends,
+			// e.g. if all terminating then all active and so on. Since we must support
+			// headless services we can't go via [Frontend.Backends] and must produce
+			// that list of backends separately.
```
Issue please

joamaki replied:
Part of https://github.com/cilium/cilium/issues/34176

dylandreimerink commented on pkg/ciliumenvoyconfig/exp_reconciler.go:
```diff
@@ -0,0 +1,149 @@
+func (ops *envoyOps) Delete(ctx context.Context, _ statedb.ReadTxn, res *EnvoyResource) error {
+	if len(res.Redirects) > 0 {
+		// Remove redirects from services no longer selected by the CEC
+		wtxn := ops.writer.WriteTxn()
```
nit: Would be good to add the `defer wtxn.Abort()`, just so its there in case someone adds an early return in the future

dylandreimerink commented on pkg/ciliumenvoyconfig/exp_tables.go:
```diff
@@ -144,3 +124,106 @@ func NewCECTable(db *statedb.DB) (statedb.RWTable[*CEC], error) {
+	return r.Status
+}
+
+func (r *EnvoyResource) Clone() *EnvoyResource {
+	r2 := *r
+	return &r2
+}
```
Is a shallow copy enough here? Since `EnvoyResource` contains maps?

joamaki replied:
Shallow copy is enough. Deep copy would be way too expensive.

dylandreimerink commented on pkg/loadbalancer/experimental/redirectpolicy/table.go:
```diff
@@ -0,0 +1,102 @@
+	lrpIDIndex = statedb.Index[*LocalRedirectPolicy, k8s.ServiceID]{
+		Name: "id",
+		FromObject: func(obj *LocalRedirectPolicy) index.KeySet {
+			// FIXME: Figure out which type to use for k8s object names. NamespacedName?
```
Is this something that needs addressing now?

joamaki replied:
Not really. More of a long term cleanup that we should do later. My thinking was we'd do perhaps a name type that is cluster+namespace+name and is just one string + indexes of the '/'s. This would avoid a lot of unnecessary string concatenations. But not strictly needed so I'll follow up with that much later.

joamaki replied:
Removed the comment. I'll get to this later! :)

joamaki committed:
experimental: ServiceCache and ServiceManager adapters. Fixes.

- Adapters for the old ServiceCache and ServiceManager APIs
- Merge BackendInstance and BackendParams. Refer to BackendParams
  from Frontend. This removes the need to call GetInstance() and
  simplifies implementation in presence of redirected services
  (where the ServiceName might be different).
- Respect the KubeProxyReplacement option
- Fall back to terminating backends when no active exist
- Bind health server to all interfaces by default
- Use bpf.MapOut for LBMaps for correct startup ordering
- Handle headless services
- Properly YAML marshal L3n4Addr

Signed-off-by: Jussi Maki <jussi@isovalent.com>
